### PR TITLE
feat(ctb): add MigrationSystemDictator checks

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Adds checks for the MigrationSystemDictator steps.

### CODE REVIEW

**Comment by mslipper** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3897** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3897" target="_blank">Graphite</a> 
    * **PR #3898** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3898" target="_blank">Graphite</a> 
      * **PR #3899** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3899" target="_blank">Graphite</a> 
        * **PR #3878** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3878" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/3878?utm_source=stack-comment).

**Review by mslipper — APPROVED** [review]

File: packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts [hunk 1]
```diff
@@ -78,88 +113,251 @@ const deployFn: DeployFunction = async (hre) => {
...
+        'proposer',
+        hre.deployConfig.l2OutputOracleProposer
+      )
+      await assertContractVariable(
+        L2OutputOracle,
+        'owner',
+        hre.deployConfig.l2OutputOracleOwner
+      )
+      if (
+        hre.deployConfig.l2OutputOracleGenesisL2Output !==
```
Comment (tynes) [inline packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts:243]: Hmm on the prod network it should not be `bytes32(0)`

File: packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts [hunk 2]
```diff
@@ -78,88 +113,251 @@ const deployFn: DeployFunction = async (hre) => {
...
+        `OptimismPortal was not initialized with the correct initial bought gas`
+      )
+      assert(
+        !resourceParams.prevBlockNum.eq(0),
+        `OptimismPortal was not initialized with the correct initial block number`
+      )
+      assert(
+        (
+          await (hre as any).ethers.provider.getBalance(OptimismPortal.address)
+        ).gt(0)
```
Comment (tynes) [inline packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts:278]: Could we get the balance at a previous step and assert the exact balance?

Comment (tynes) [reply packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts:278]: I suppose the 0 balance check on the bridge sort of covers this

**Review by tynes — COMMENTED** [review]
Mostly small things, generally looks good to me

File: packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts (hunk 1 above)
Comment (smartcontracts) [reply packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts:245]: Added a big warning for this

File: packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts [hunk 3]
```diff
@@ -78,88 +113,251 @@ const deployFn: DeployFunction = async (hre) => {
...
+        resourceParams.prevBoughtGas.eq(0),
+        `OptimismPortal was not initialized with the correct initial bought gas`
+      )
+      assert(
+        !resourceParams.prevBlockNum.eq(0),
+        `OptimismPortal was not initialized with the correct initial block number`
+      )
+      assert(
+        (
+          await (hre as any).ethers.provider.getBalance(OptimismPortal.address)
```
[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts]: Done

File: packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts [hunk 4]
```diff
@@ -1,74 +1,109 @@
...
-        `WARNING!!! Make sure you are ONLY doing this on a test network.`
-      )
-      controller = deployer
-      isLiveDeployer = true
-    } else {
-      throw new Error(
-        `controller address MUST NOT be the deployer on live networks`
-      )
-    }
+  if (hre.deployConfig.controller === deployer) {
```
[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts-bedrock/deploy/018-MigrationSystemDictatorSteps.ts:20]: Done

**Review by tynes — APPROVED** [review]

**Pull request merged**